internal/component/pyroscope/ebpf/send.go

```diff
@@ -0,0 +1,51 @@
+//go:build (linux && arm64) || (linux && amd64)
+
+package ebpf
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	"github.com/go-kit/log/level"
+	"github.com/grafana/alloy/internal/component/pyroscope"
+	"github.com/grafana/pyroscope/ebpf/pprof"
+)
+
+const maxSendConcurrency = 32
+
+func (c *Component) sendProfiles(ctx context.Context, builders *pprof.ProfileBuilders) {
+	pool := workerPool{}
+	pool.run(min(maxSendConcurrency, len(builders.Builders)), len(builders.Builders))
+	defer func() {
+		pool.stop()
+	}()
+	for _, builder := range builders.Builders {
+		if ctx.Err() != nil {
+			return
+		}
+
+		serviceName := builder.Labels.Get("service_name")
+		c.metrics.pprofsTotal.WithLabelValues(serviceName).Inc()
+		c.metrics.pprofSamplesTotal.WithLabelValues(serviceName).Add(float64(len(builder.Profile.Sample)))
+
+		buf := bytes.NewBuffer(nil)
+		_, err := builder.Write(buf)
+		if err != nil {
+			level.Error(c.options.Logger).Log("err", fmt.Errorf("ebpf profile encode %w", err))
+			continue
+		}
+		rawProfile := buf.Bytes()
+
+		appender := c.appendable.Appender()
+		c.metrics.pprofBytesTotal.WithLabelValues(serviceName).Add(float64(len(rawProfile)))
+
+		pool.do(func() {
+			samples := []*pyroscope.RawSample{{RawProfile: rawProfile}}
+			err = appender.Append(ctx, builder.Labels, samples)
+			if err != nil {
+				level.Error(c.options.Logger).Log("msg", "ebpf pprof write", "err", err)
+			}
+		})
```

Comment:
`pool.do` is blocking and `appender.Append` may take a really long time in practice. Should we maybe do something like:
```go
// The timeout should be aligned with the collection interval:
// if we didn't manage to send everything, we won't block profiler (if I get it right)
ctx, cancel := context.WithTimeout(ctx, timeout)
defer cancel()

for (something) {
   // IDK if we're allowed to call it concurrently
   c.appendable.Appender()
   job := func() {
      err = appender.Append(ctx, builder.Labels, samples)
      // Log
   }
   select {
   case pool.jobs <- job:
   case <-ctx.Done():
      // Log
      return
   }
}

```

Reply:
I think there may be issue if I just do `defer cancel()` it will cancel the context for the last 32 profiles batch.

I think I better do 
```go
	defer func() {
		pool.stop() // wait for profiles to be sent and goroutines to finish. This should be quick if the timeout deadline already reached assuming the fanoutClient respects the ctx
		cancel() // release the timer if we sent profiles faster then timeout
	}()
```

Reply:
Yeah, I didn't test that, just wanted to avoid blocking on `pool.do()`